### PR TITLE
[NFC] Use unique_ptr in SparseSet

### PR DESCRIPTION
**Pull request by optimisan**
This allows implementing the move constructor.

### CODE REVIEW

**Comment by optimisan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116617?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116618** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116618?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116617** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116617?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116617?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116616** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116616?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: llvm/include/llvm/ADT/SparseSet.h [hunk 1]
```diff
@@ -205,7 +209,7 @@ class SparseSet {
     assert(Idx < Universe && "Key out of range");
     assert(Sparse != nullptr && "Invalid sparse type");
     const unsigned Stride = std::numeric_limits<SparseT>::max() + 1u;
-    for (unsigned i = Sparse[Idx], e = size(); i < e; i += Stride) {
+    for (unsigned i = Sparse.get()[Idx], e = size(); i < e; i += Stride) {
```
Comment (dwblaikie) [inline llvm/include/llvm/ADT/SparseSet.h:212]: If you make the `std::unique_ptr<SparseT, Deleter>` into a `std::unique_ptr<SparseT[], Deleter>` then you can use [] directly without the `.get()` I think?

**Comment by dwblaikie** [comment]
Oh, and please add unit test coverage for the new move functionality.

**Review by dwblaikie — APPROVED** [review]
LGTM - thanks!

**Comment by optimisan** [comment]
### Merge activity

* **Nov 22, 6:17 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/116617).

**Pull request merged**